### PR TITLE
config: add option to configure IP addresses through command line

### PR DESCRIPTION
**Pull request by redbeam**
Adds command line arguments to configure subnet, gateway IP, device IP, host IP, with default values. Specifying only the subnet without other IPs will generate appropriate IPs for use.
Also adds new entries to config.yaml (`deviceIP`, `hostIP`).

### CODE REVIEW

**Comment by cfergeau** [comment]
## Code Review: PR #614 - Add IP Address Configuration Options

### Overview
This PR adds command-line arguments and config file options to configure network IP addresses (subnet, gateway, device, and host IPs). Previously, these were hardcoded to `192.168.127.0/24` with fixed offsets. The change enables users to customize their network topology while maintaining backward compatibility through sensible defaults.

**Changes:**
- +232 lines / -45 lines
- New CLI flags: `--subnet`, `--gatewayIP`, `--deviceIP`, `--hostIP`
- New config fields: `deviceIP`, `hostIP` in `types.Configuration`
- Refactored IP calculation logic with validation
- Comprehensive test updates

---

### Code Quality & Correctness

#### ✅ Strengths

1. **Backward Compatibility**: Default behavior is preserved - existing users won't notice any changes
2. **Comprehensive Testing**: All test cases updated with new fields, including a test for omitted deviceIP/hostIP
3. **Validation**: IP addresses are validated for correctness and subnet membership
4. **Clear Error Messages**: Helpful error messages when IPs are invalid or outside subnet
5. **Refactoring**: Removed hardcoded constants, making code more maintainable

#### ⚠️ Issues & Suggestions

**1. Flag Naming Convention (Minor)**
```go
flagSet.StringVar(&args.gatewayIP, "gatewayIP", "", "...")
flagSet.StringVar(&args.deviceIP, "deviceIP", "", "...")
flagSet.StringVar(&args.hostIP, "hostIP", "", "...")
```
- Go CLI tools typically use kebab-case: `--gateway-ip`, `--device-ip`, `--host-ip`
- Current camelCase is inconsistent with existing flags like `--ssh-port`, `--config-file`
- **Recommendation**: Change to kebab-case for consistency

**2. Silent Error Suppression (Bug Risk)**
```go
func subnetContains(subnet netip.Prefix, checkIP string) bool {
	ip, _ := netip.ParseAddr(checkIP)  // Error ignored!
	return subnet.Contains(ip)
}
```
- The comment says "expects that the checkIP address is already valid", but there's no guarantee
- If `ParseAddr` fails, `ip` is zero value and `Contains` may give wrong result
- **Recommendation**: Either panic on error (since it's a precondition violation) or return `(bool, error)`

**3. IP Conflict Validation Missing (Important)**

The code validates that IPs are in the subnet but doesn't check if they conflict:
```go
// What if gatewayIP == deviceIP?
// What if hostIP == deviceIP?
```
- **Recommendation**: Add validation that the three IPs are distinct

**4. Redundant Helper Function**
```go
func isValidIP(ip string) bool {
	_, err := netip.ParseAddr(ip)
	return err == nil
}
```
This is only used once in `validateConfigIPs` and could be inlined for better error messages.

**5. Network vs Broadcast Address Validation**

The code doesn't prevent using the network or broadcast address:
```go
// This would be allowed but shouldn't:
// --subnet 192.168.1.0/24 --gatewayIP 192.168.1.0     (network address)
// --subnet 192.168.1.0/24 --hostIP 192.168.1.255      (broadcast address)
```
- **Recommendation**: Add validation to reject network/broadcast addresses

**6. Error Message Enhancement**
```go
return config, fmt.Errorf("gateway IP (%s) not in subnet (%s)", config.Stack.GatewayIP, subnet)
```
Could be more helpful:
```go
return config, fmt.Errorf("gateway IP %s is outside subnet %s (usable range: %s - %s)", 
    config.Stack.GatewayIP, subnet, firstUsable, lastUsable)
```

---

### Test Coverage

#### ✅ Good Coverage
- Test for legacy behavior without new flags
- Test for custom subnet with automatic IP generation
- Test for explicit IP configuration
- Test for backward compatibility (config without deviceIP/hostIP)

#### ❌ Missing Test Cases

1. **Negative test**: IPs outside subnet should fail
2. **Negative test**: Duplicate IPs (gateway == device) should fail
3. **Negative test**: Using network/broadcast addresses should fail
4. **Edge case**: Smallest viable subnet (/29) with all IPs specified
5. **IPv6 test**: Verify IPv6 subnet configuration works end-to-end

---

### Suggested Code Changes

Here's a suggested fix for the `subnetContains` function:

```go
// Before (unsafe):
func subnetContains(subnet netip.Prefix, checkIP string) bool {
	ip, _ := netip.ParseAddr(checkIP)  // Silently ignores error!
	return subnet.Contains(ip)
}

// After (safe):
func subnetContains(subnet netip.Prefix, checkIP string) bool {
	ip, err := netip.ParseAddr(checkIP)
	if err != nil {
		// This should never happen since validateConfigIPs runs first,
		// but defensive programming is good
		panic(fmt.Sprintf("BUG: invalid IP passed to subnetContains: %s", checkIP))
	}
	return subnet.Contains(ip)
}
```

And add IP conflict validation:

```go
func validateConfigIPs(config types.Configuration) error {
	if _, err := netip.ParsePrefix(config.Subnet); err != nil {
		return fmt.Errorf("failed to parse subnet: %w", err)
	}

	// Validate IP address format
	ips := map[string]string{
		"gateway": config.GatewayIP,
		"device":  config.DeviceIP,
		"host":    config.HostIP,
	}
	for name, ip := range ips {
		if _, err := netip.ParseAddr(ip); err != nil {
			return fmt.Errorf("invalid %s IP address %q: %w", name, ip, err)
		}
	}

	// Check for duplicate IPs
	if config.GatewayIP == config.DeviceIP {
		return fmt.Errorf("gateway IP and device IP must be different (both are %s)", config.GatewayIP)
	}
	if config.GatewayIP == config.HostIP {
		return fmt.Errorf("gateway IP and host IP must be different (both are %s)", config.GatewayIP)
	}
	if config.DeviceIP == config.HostIP {
		return fmt.Errorf("device IP and host IP must be different (both are %s)", config.DeviceIP)
	}

	return nil
}
```

---

### Overall Assessment

**Quality**: Good - well-structured, well-tested, maintains backward compatibility

**Risk Level**: Low-Medium - the silent error in `subnetContains()` and missing duplicate IP validation are the main concerns

**Recommendation**: Request changes to address the critical issues (silent error, duplicate IP validation, flag naming), then approve after fixes.

---

### Recommendations Summary

**Must Fix:**
1. Fix silent error in `subnetContains()` - this could cause bugs
2. Add validation for duplicate IPs
3. Use kebab-case for flag names (`--gateway-ip`, etc.) for consistency

**Should Fix:**
4. Add validation to reject network/broadcast addresses
5. Add negative test cases for error conditions
6. Improve error messages with usable IP range

**Nice to Have:**
7. Inline `isValidIP()` helper for better error messages
8. Add IPv6 integration test
9. Add comment in example config that deviceIP/hostIP are optional

**Comment by cfergeau** [comment]
> Code Review: PR #614 - Add IP Address Configuration Options

I’ve asked claude to post this as I agree with most of the comments, it’s just the "must fix"/"should fix"/"nice to have" classification which I’m not sure about :)
Happy to discuss if you think some of these changes are not useful.

**Comment by redbeam** [comment]
@cfergeau Here are my comments:

```
1. I'll go with whatever you prefer @cfergeau 
2. I'll add a panic if the input is invalid.
4. I added this as I think it makes the code a bit cleaner.
6. I think the error message is fine as is.

3 & 5. Good points, will fix.
```

**Comment by cfergeau** [comment]
> @cfergeau Here are my comments:
> 
> ```
> 1. I'll go with whatever you prefer @cfergeau 
> 2. I'll add a panic if the input is invalid.
> 4. I added this as I think it makes the code a bit cleaner.
> 6. I think the error message is fine as is.
> 
> 3 & 5. Good points, will fix.
> ```

For 1., you could remove the helper, or maybe do this:
```
diff --git a/cmd/gvproxy/config.go b/cmd/gvproxy/config.go
index 5e247dc0..392416a1 100644
--- a/cmd/gvproxy/config.go
+++ b/cmd/gvproxy/config.go
@@ -241,14 +241,14 @@ func GvproxyConfigure(config *GvproxyConfig, args *GvproxyArgs, version string)
 	}
 
 	// Check whether configured IPs are inside the subnet
-	if !subnetContains(subnet, config.Stack.GatewayIP) {
-		return config, fmt.Errorf("gateway IP (%s) not in subnet (%s)", config.Stack.GatewayIP, subnet)
+	if err := subnetContains(subnet, config.Stack.GatewayIP); err != nil {
+		return config, err
 	}
-	if !subnetContains(subnet, config.Stack.DeviceIP) {
-		return config, fmt.Errorf("device IP (%s) not in subnet (%s)", config.Stack.DeviceIP, subnet)
+	if err := subnetContains(subnet, config.Stack.DeviceIP) ; err != nil {
+		return config, err
 	}
-	if !subnetContains(subnet, config.Stack.HostIP) {
-		return config, fmt.Errorf("host IP (%s) not in subnet (%s)", config.Stack.HostIP, subnet)
+	if err := subnetContains(subnet, config.Stack.HostIP); err != nil {
+		return config, err
 	}
 
 	if config.Stack.GatewayMacAddress == "" {
@@ -515,8 +515,14 @@ func validateConfigIPs(config types.Configuration) error {
 	return nil
 }
 
-// This function expects that the checkIP address is already valid.
-func subnetContains(subnet netip.Prefix, checkIP string) bool {
-	ip, _ := netip.ParseAddr(checkIP)
-	return subnet.Contains(ip)
+func subnetContains(subnet netip.Prefix, checkIP string)(error) {
+	ip, err := netip.ParseAddr(checkIP)
+	if err != nil {
+		return err
+	}
+	if !subnet.Contains(ip) {
+		return fmt.Errorf("IP (%s) not in subnet (%s)", checkIP, subnet)
+	}
+
+	return nil
 }
```

While I agree in this case, not having the error is not a problem, I usually end up having to add the `error` arg when I refactor the code/want to use the helper some place else/…

**Comment by redbeam** [comment]
@cfergeau 
Note: I referred to the points in "⚠️ Issues & Suggestions" section, not "Recommendations Summary" 😅

Anyway, I refactored the validation function. It checks for multiple conditions regarding the IPs.

**Comment by cfergeau** [comment]
Can’t get `go test` to get green, but I suspect this is the usual flakeyness we are seeing with these tests :(
I tried it locally and it passed.
/lgtm
/approve
